File: src/main/kotlin/io/titandata/titan/providers/local/Uninstall.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.local

import io.titandata.client.apis.RepositoriesApi
import io.titandata.titan.clients.Docker
import io.titandata.titan.exceptions.CommandException
import io.titandata.titan.utils.CommandExecutor
import io.titandata.titan.utils.ProgressTracker

class Uninstall (
    private val titanServerVersion: String,
    private val exit: (message: String, code: Int) -> Unit,
    private val remove: (container: String, force: Boolean) -> Unit,
    private val commandExecutor: CommandExecutor = CommandExecutor(),
    private val docker: Docker = Docker(commandExecutor),
    private val repositoriesApi: RepositoriesApi = RepositoriesApi(),
    private val track: (title: String, function: () -> Any) -> Unit = ProgressTracker()::track
) {
    fun uninstall(force: Boolean) {
        if (docker.titanServerIsAvailable()) {
            val repositories = repositoriesApi.listRepositories()
            for (repo in repositories) {
                if (!force) {
                    exit("repository ${repo.name} exists, remove first or use '-f'", 1)
                }
                track("Removing repository ${repo.name}" ) { remove(repo.name, true) }
            }
        }
        if (docker.titanServerIsAvailable()) docker.rm("titan-server", true)
        if (docker.titanLaunchIsAvailable()) docker.rm("titan-launch", true)
        track ("Tearing down Titan servers") {
            docker.teardownTitanServers()
        }
        track ("Removing titan-data Docker volume") {
            docker.removeVolume("titan-data")
        }
        track ("Removing Titan Docker image") {
            docker.removeTitanImages(titanServerVersion)
        }
        println("Uninstalled titan infrastructure")
    }
}

File: internal/app/providers/local/Uninstall.go
package local

import (
	"fmt"
	"os"
	"strconv"
	"titan/internal/app/clients"
)

func Uninstall(version string, force bool, removeImages bool, port int, context string) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)
	docker := clients.Docker(context, port)

	serverAvailable, _ := docker.TitanServerIsAvailable()
	if serverAvailable {
		var repos, _, _ = repositoriesApi.ListRepositories(ctx)
		for _, repo := range repos {
			if !force {
				fmt.Println("repository '" + repo.Name + "' exists, remove first or use '-f'")
				os.Exit(1)
			}
			Remove(repo.Name, true, port, context)
		}
	}
	if serverAvailable {
		s, err := docker.RemoveTitanServer()
		if err != nil {
			fmt.Println(s)
			panic(err)
		}
	}
	launchAvailable, _ := docker.TitanLaunchIsAvailable()
	if launchAvailable {
		s, err := docker.RemoveTitanLaunch()
		if err != nil {
			fmt.Println(s)
			panic(err)
		}
	}

	fmt.Println("Tearing down Titan servers")
	docker.TeardownTitanServers() //TODO track this

	fmt.Println("Removing titan-data Docker volume")
	docker.RemoveTitanVolume() //TODO track this

	if removeImages {
		fmt.Println("Removing Titan Docker image")
		docker.RemoveTitanImages(version) //TODO track this
	}
	fmt.Println("Uninstalled titan infrastructure")
}